packages/rendermime/src/renderers.ts

```diff
@@ -511,38 +522,94 @@ function autolink(content: string): Array<HTMLAnchorElement | Text> {
-        document.createTextNode(content.slice(lastIndex, match.index))
-      );
+  const { checkPathUrls, checkWebUrls } = options;
+
+  const regexes = [webLinkRegex, pathLinkRegex];
+  const enabled = [checkWebUrls, checkPathUrls];
+  const kinds = ['web', 'path'];
+
+  const nodes: Array<HTMLAnchorElement | Text> = [];
+
+  const linkify = (content: string, regexIndex: number) => {
+    if (regexIndex >= regexes.length) {
+      nodes.push(document.createTextNode(content));
+      return;
     }
-    let url = match[0];
-    // Special case when the URL ends with ">" or "<"
-    const lastChars = url.slice(-1);
-    const endsWithGtLt = ['>', '<'].indexOf(lastChars) !== -1;
-    const len = endsWithGtLt ? url.length - 1 : url.length;
-    const anchor = document.createElement('a');
-    url = url.slice(0, len);
-    anchor.href = url.startsWith('www.') ? 'https://' + url : url;
-    anchor.rel = 'noopener';
-    anchor.target = '_blank';
-    anchor.appendChild(document.createTextNode(url.slice(0, len)));
-    nodes.push(anchor);
-    lastIndex = match.index + len;
-  }
-  if (lastIndex !== content.length) {
-    nodes.push(
-      document.createTextNode(content.slice(lastIndex, content.length))
-    );
-  }
+
+    const regex = regexes[regexIndex];
+    const isEnabled = enabled[regexIndex];
+
+    if (isEnabled) {
+      let match;
+      let currentIndex = 0;
+      regex.lastIndex = 0;
+
+      while ((match = regex.exec(content)) !== null) {
+        const stringBeforeMatch = content.substring(currentIndex, match.index);
+        if (stringBeforeMatch) {
+          linkify(stringBeforeMatch, regexIndex + 1);
+        }
+
+        const value = match[0];
+        if (kinds[regexIndex] === 'web') {
+          nodes.push(convertToUrlAnchor(value));
+        } else if (kinds[regexIndex] === 'path') {
+          nodes.push(convertToPathAnchor(value));
+        }
+
+        currentIndex = match.index + value.length;
+      }
+      const stringAfterMatches = content.substring(currentIndex);
+      if (stringAfterMatches) {
+        linkify(stringAfterMatches, regexIndex + 1);
+      }
+    } else {
+      linkify(content, regexIndex + 1);
+    }
+  };
+
+  linkify(content, 0);
   return nodes;
 }
 
+function convertToUrlAnchor(url: string): HTMLAnchorElement {
+  // Special case when the URL ends with ">" or "<"
+  const lastChars = url.slice(-1);
+  const endsWithGtLt = ['>', '<'].indexOf(lastChars) !== -1;
+  const len = endsWithGtLt ? url.length - 1 : url.length;
+  const anchor = document.createElement('a');
+  url = url.slice(0, len);
+  anchor.href = url.startsWith('www.') ? 'https://' + url : url;
+  anchor.rel = 'noopener';
+  anchor.target = '_blank';
+  anchor.appendChild(document.createTextNode(url.slice(0, len)));
+  return anchor;
+}
+
+function convertToPathAnchor(path: string): HTMLAnchorElement {
+  const anchor = document.createElement('a');
+  anchor.innerText = path.slice(0, path.length);
+  anchor.href = path;
```

Comment:
UX-wise this may lead to a number of "broken" links because:
- the path may be outside of the JupyterLab root
- there will be false positive matches

One solution would be to initially render the path anchors as `<span>`s which later asynchronously get swapped with anchors after checking that a file under such path can be opened.